doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,940 @@
+to our Parquet files in the form of a [protobuf message](https://github.com/MaterializeInc/materialize/blob/7135e53182c77e2c2fe31fe2aa700adc37ac134d/src/persist/src/persist.proto#L26-L35)
+I propose we extend this message to include include a "version" which initially
+will just be `1` and can be incremented at its own pace. The goal of this
+"batch version" is it gives us a way to entirely ignore the new columnar data
+if necessary. For example, say we incorrectly encode the new structured format
+and should always ignore the bad data on decode, bumping the version number
+gives us an easy way to do that.
+
+### Migration
+
+#### Introducing Columnar/Structured Data
+
+Today data is encoded with the `Codec` trait and written to S3 with the
+following Parquet schema:
+
+```
+{
+  "k": bytes,
+  "v": bytes,
+  "t": i64,
+  "d": i64,
+}
+```
+
+To migrate to the new structured data I propose we extend the above schema to:
+
+```
+{
+  "k": bytes,
+  "v": bytes,
+  "t": i64,
+  "d": i64,
+  "k_s": {
+    // ... nested structured data
+  },
+  "v_s": {
+    // ... nested structured data
+  },
+}
+```
+
+In other words, we dual encode blobs with both `Codec` and Arrow and write them
+to a single Parquet file. This has the following benefits:
+
+1. Slow rollout/shadow migration. If we write both formats we can let this
+   change bake for as long as we would like, validating that `Codec` and Arrow
+   decode to the same `Row`s. Reading only Arrow data could be turned on or off
+   via a LaunchDarkly flag.
+2. Incremental progress. If we write both formats then we could read back Arrow
+   data but sort and consolidate with the `Codec` format. Currently decoding
+   protobuf (i.e. `ProtoDatum`) is computationally expensive, getting an
+   incremental win of improving decode time would be nice.
+3. Doesn't increase the number of `PUT` or `GET` requests to `S3`.
+4. Doesn't require any change to state stored in Consensus.
+5. (Forward looking) When we switch to reading only Arrow data we can leverage
+   Parquet's projection pushdown to ignore the old `'k'` and `'v'` columns.
+
+The current max blob size is 128MiB, we have two options for how we want to
+handle the "max" size when we write both `Codec` data and Arrow at the same
+time.
+
+1. Ignore Arrow when calculating the max. This increases the size of S3 objects
+   which may impact performance for things that have been tuned to assume
+   128MiB sized objets, e.g. compute back pressure.
+2. Include Arrow when calculating the max. This reduces the amount of rows that
+   can get stored in a single blob, thus increasing the number of Parts
+   per-Run. This increases the number of S3 `GET`s and `PUT`s which could
+   impact cost, and requires `ConsolidatingIter` to handle a larger number of
+   Parts.
+
+Neither option is great, but I am leanning towards [1] since it puts the
```

Comment:
> Another big factor is memory pressure.

Very true! Added a note about Prometheus metrics to track the size of blobs we read from S3.

> Maybe we can get some of this back by enabling parquet compression on all of k, v, k_s, v_s

This would be great! Parquet supports a number of compression types, I'll need to read up but I think either LZ4 or Snapp would be good here? AFAIU they're fairly fast compression algos. Another win we could get is enabling run-length encoding on the `diff` and `time` columns. Thinking about some of our largest blobs which are physically compacted (I think?) all of the time and diff fields should be the same(?).